### PR TITLE
Log4j 2 Programmatic Configuration BAEL 1338

### PR DESCRIPTION
**Pull request by ShyamVeda**
Log4j 2 Programmatic Configuration - BAEL 1338 by v.shyamsundar1988@gmail.com

### CODE REVIEW

File: logging-modules/log4j2-programmatic-configuration/pom.xml [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+	<packaging>pom</packaging>
+	<modules>
+		<module>simple-configuration</module>
+		<module>set-configuration-factory</module>
+		<module>simple-configurator</module>
+		<module>modifying-existing-config</module>
+		<module>simple-configuration-xml</module>
+		<module>custom-appender</module>
+		<module>custom-appenders</module>
+    <module>modify-xml-configuration</module>
```
Comment (jzheaux) [inline logging-modules/log4j2-programmatic-configuration/pom.xml]: Let's fix the whitespace issues here. Would prefer spaces over tabs in the file.

File: logging-modules/log4j2-programmatic-configuration/set-configuration-factory/src/main/java/com/baeldung/log4j2/config/CustomConfigurationFactory.java [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+            .addComponent(builder.newComponent("SizeBasedTriggeringPolicy")
+                .addAttribute("size", "100M"));
+        AppenderComponentBuilder rollingFile = builder.newAppender("rolling", "RollingFile");
+        rollingFile.addAttribute("fileName", "target/rolling.log");
+        rollingFile.addAttribute("filePattern", "target/archive/rolling-%d{MM-dd-yy}.log.gz");
+        rollingFile.add(layout);
+        rollingFile.addComponent(triggeringPolicy);
+        builder.add(rollingFile);
+
+        /*AppenderComponentBuilder syslog = builder.newAppender("syslogAppender", "Syslog");
```
Comment (jzheaux) [inline logging-modules/log4j2-programmatic-configuration/set-configuration-factory/src/main/java/com/baeldung/log4j2/config/CustomConfigurationFactory.java]: Let's clean this up. Either leave it in, take it out, or move it to another example. Same with the other comments below.

File: logging-modules/log4j2-programmatic-configuration/set-configuration-factory/src/test/java/com/baeldung/log4j2/logtest/LogTest.java [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+
+import com.baeldung.log4j2.config.CustomConfigurationFactory;
+
+@RunWith(JUnit4.class)
+public class LogTest {
+    static {
+        CustomConfigurationFactory customConfigurationFactory = new CustomConfigurationFactory();
+        ConfigurationFactory.setConfigurationFactory(customConfigurationFactory);
+    }
+    @Test
```
Comment (jzheaux) [inline logging-modules/log4j2-programmatic-configuration/set-configuration-factory/src/test/java/com/baeldung/log4j2/logtest/LogTest.java]: Please separate the static config and the @Test with a blank line

File: logging-modules/log4j2-programmatic-configuration/simple-configuration-xml/src/test/java/com/baeldung/log4j2/logtest/LogTest.java [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+
+@RunWith(JUnit4.class)
+public class LogTest {
+
+    @Test
+    public void simpleProgrammaticConfiguration(){
```
Comment (jzheaux) [inline logging-modules/log4j2-programmatic-configuration/simple-configuration-xml/src/test/java/com/baeldung/log4j2/logtest/LogTest.java]: Please put a space between the () and the {.

File: logging-modules/log4j2-programmatic-configuration/modify-xml-configuration/src/main/resources/log4j2.xml [hunk 5]
```diff
@@ -0,0 +1,16 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<Configuration monitorInterval="60">
+    <Appenders>
+        <Console name="console" target="SYSTEM_OUT">
+            <PatternLayout pattern="[%-5level] %d{yyyy-MM-dd HH:mm:ss.SSS} [%t] %c{1} - %msg%n"/>
+        </Console>
+    </Appenders>
+    <Loggers>
+    	<Logger name="com" additivity="false" level="info">
```
Comment (jzheaux) [inline logging-modules/log4j2-programmatic-configuration/modify-xml-configuration/src/main/resources/log4j2.xml]: Let's make the indentation consistent. Seems that in some places in this file, it is 2 spaces while in others it is 4.

File: logging-modules/log4j2-programmatic-configuration/modify-xml-configuration/pom.xml [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+  <name>modify-xml-configuration</name>
+  <url>http://maven.apache.org</url>
+  <properties>
+    <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
+  </properties>
+  <dependencies>
+    <dependency>
+      <groupId>junit</groupId>
+      <artifactId>junit</artifactId>
+      <version>3.8.1</version>
```
Comment (jzheaux) [inline logging-modules/log4j2-programmatic-configuration/modify-xml-configuration/pom.xml]: Let's use JUnit 4.

File: logging-modules/log4j2-programmatic-configuration/pom.xml [hunk 7]
```diff
@@ -0,0 +1,36 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+	<modelVersion>4.0.0</modelVersion>
+	<groupId>com.baeldung.log4j2</groupId>
+	<artifactId>log4j2-programmatic-configuration</artifactId>
```
Comment (jzheaux) [inline logging-modules/log4j2-programmatic-configuration/pom.xml]: Let's make sure that this is added to the main build. I think you will need to modify the [main pom](https://github.com/eugenp/tutorials/blob/master/pom.xml). Just grep for `logging-modules`.

File: logging-modules/log4j2-programmatic-configuration/set-configuration-factory/src/main/java/com/baeldung/log4j2/config/CustomConfigurationFactory.java [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+import org.apache.logging.log4j.core.config.builder.api.AppenderComponentBuilder;
+import org.apache.logging.log4j.core.config.builder.api.ComponentBuilder;
+import org.apache.logging.log4j.core.config.builder.api.ConfigurationBuilder;
+import org.apache.logging.log4j.core.config.builder.api.FilterComponentBuilder;
+import org.apache.logging.log4j.core.config.builder.api.LayoutComponentBuilder;
+import org.apache.logging.log4j.core.config.builder.api.LoggerComponentBuilder;
+import org.apache.logging.log4j.core.config.builder.api.RootLoggerComponentBuilder;
+import org.apache.logging.log4j.core.config.builder.impl.BuiltConfiguration;
+
+public class CustomConfigurationFactory extends ConfigurationFactory {
```
Comment (jzheaux) [inline logging-modules/log4j2-programmatic-configuration/set-configuration-factory/src/main/java/com/baeldung/log4j2/config/CustomConfigurationFactory.java]: There is a lot of code to soak in here on first read. Let's add a header to each of these configurations which explains what each class is accomplishing that is different from the other examples.

File: logging-modules/log4j2-programmatic-configuration/simple-configuration/src/main/java/com/baeldung/log4j2/config/CustomConfigurationFactory.java [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+        LayoutComponentBuilder layout = builder.newLayout("PatternLayout")
+            .addAttribute("pattern", "%d [%t] %-5level: %msg%n%throwable");
+        console.add(layout);
+        FilterComponentBuilder filter = builder.newFilter("MarkerFilter", Filter.Result.ACCEPT, Filter.Result.DENY);
+        filter.addAttribute("marker", "FLOW");
+        console.add(filter);
+        builder.add(console);
+
+        ComponentBuilder triggeringPolicy = builder.newComponent("Policies")
+            .addComponent(builder.newComponent("CronTriggeringPolicy")
```
Comment (jzheaux) [inline logging-modules/log4j2-programmatic-configuration/simple-configuration/src/main/java/com/baeldung/log4j2/config/CustomConfigurationFactory.java]: On baeldung, the whitespace standard is 2 spaces for continuations and 4 for blocks. You can set these in your editor, and then you won't have to remember.

File: logging-modules/log4j2-programmatic-configuration/simple-configurator/src/test/java/com/baeldung/log4j2/logtest/LogPrinter.java [hunk 10]
```diff
@@ -0,0 +1,15 @@
+package com.baeldung.log4j2.logtest;
+
+import org.apache.logging.log4j.core.Logger;
+import org.apache.logging.log4j.core.LoggerContext;
+
+public class LogPrinter {
+    private Logger logger = LoggerContext.getContext(false)
```
Comment (jzheaux) [inline logging-modules/log4j2-programmatic-configuration/simple-configurator/src/test/java/com/baeldung/log4j2/logtest/LogPrinter.java]: OOC, why not just do `LogManager.getLogger`?

**Comment by jzheaux** [comment]
Nice code, overall. Most of the changes are to bring it in line with the site's whitespace standards. I'd recommend changing your editor to use the standard so that you don't have to remember.

Also, I ended up changing some of the code in the article one last time, just to make things fit better on the page. You'll notice that some variable and type names I made shorter. Would you please change the code here to bring that in line? Thank you.

Looks really good, @ShyamVeda.

File: logging-modules/log4j2-programmatic-configuration/set-configuration-factory/src/main/java/com/baeldung/log4j2/config/CustomConfigurationFactory.java (hunk 8 above)
Comment (ShyamVeda) [reply logging-modules/log4j2-programmatic-configuration/set-configuration-factory/src/main/java/com/baeldung/log4j2/config/CustomConfigurationFactory.java]: Hello @jzheaux 
Could you kindly elaborate more on this?

Thanks,
Shyam.

File: logging-modules/log4j2-programmatic-configuration/.gitignore [hunk 11]
```diff
@@ -0,0 +1,23 @@
+*.class
```
Comment (jzheaux) [inline logging-modules/log4j2-programmatic-configuration/.gitignore]: Is this necessary? The repo should already have a .gitignore.

**Pull request merged**